src/Domain/FunctionsOfTime/FunctionOfTimeHelpers.cpp

```diff
@@ -0,0 +1,117 @@
+
+#include <pup.h>
+#include <pup_stl.h>
+
+#include "Utilities/EqualWithinRoundoff.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace FunctionOfTimeHelpers {
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+StoredInfo<MaxDerivPlusOne, StoreCoefs>::StoredInfo(
+    const double t,
+    std::array<DataVector, MaxDerivPlusOne> init_quantities) noexcept
+    : time(t), stored_quantities{std::move(init_quantities)} {
+  // Convert derivs to coefficients for polynomial evaluation
+  // The coefficient of x^N is the Nth deriv rescaled by 1/factorial(N)
+  double fact = 1.0;
+  for (size_t j = 2; j < MaxDerivPlusOne; j++) {
+    fact *= j;
+    gsl::at(stored_quantities, j) /= fact;
+  }
+}
+
+template <size_t MaxDerivPlusOne>
+StoredInfo<MaxDerivPlusOne, false>::StoredInfo(
+    const double t,
+    std::array<DataVector, MaxDerivPlusOne> init_quantities) noexcept
+    : time(t), stored_quantities{std::move(init_quantities)} {}
+
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+void StoredInfo<MaxDerivPlusOne, StoreCoefs>::pup(PUP::er& p) noexcept {
+  p | time;
+  p | stored_quantities;
+}
+
+template <size_t MaxDerivPlusOne>
+void StoredInfo<MaxDerivPlusOne, false>::pup(PUP::er& p) noexcept {
+  p | time;
+  p | stored_quantities;
+}
+
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+bool StoredInfo<MaxDerivPlusOne, StoreCoefs>::operator==(
+    const StoredInfo<MaxDerivPlusOne, StoreCoefs>& rhs) const noexcept {
+  return time == rhs.time and stored_quantities == rhs.stored_quantities;
+}
+
+void reset_expiration_time(const gsl::not_null<double*> prev_expiration_time,
+                           const double next_expiration_time) noexcept {
+  if (next_expiration_time < *prev_expiration_time) {
+    ERROR("Attempted to change expiration time to "
+          << next_expiration_time
+          << ", which precedes the previous expiration time of "
+          << *prev_expiration_time << ".");
+  }
+  *prev_expiration_time = next_expiration_time;
+}
+
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+const StoredInfo<MaxDerivPlusOne, StoreCoefs>& stored_info_from_upper_bound(
+    const double t, const std::vector<StoredInfo<MaxDerivPlusOne, StoreCoefs>>&
+                        all_stored_infos) noexcept {
+  // this function assumes that the times in stored_info_at_update_times is
+  // sorted, which is enforced by the update function of a piecewise polynomial.
+
+  const auto upper_bound_stored_info = std::upper_bound(
+      all_stored_infos.begin(), all_stored_infos.end(), t,
+      [](double t0, const StoredInfo<MaxDerivPlusOne, StoreCoefs>& d) {
+        return d.time > t0;
+      });
+
+  if (upper_bound_stored_info == all_stored_infos.begin()) {
+    // all elements of times are greater than t
+    // check if t is just less than the min element by roundoff
+    if (not equal_within_roundoff(upper_bound_stored_info->time, t)) {
```

Comment:
Could `all_stored_infos` be empty?

Reply:
No it can't. The constructor of a PiecewisePolynomial has to take a time and an initial function/derivs. These are immediately `std::move`d to the first element of `all_stored_infos`

Reply:
Well actually it could be default constructed, which in that case `all_stored_infos` would be empty. I can add a check to make sure it's not empty and throw an error if it is empty.